src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt

```diff
@@ -0,0 +1,90 @@
+import jakarta.ws.rs.core.Response
+import jakarta.ws.rs.ext.Provider
+import org.eclipse.microprofile.config.inject.ConfigProperty
+
+
+@Provider
+@ApplicationScoped
+class AuthenticationFilter @Inject constructor(
+    @Context private val resourceInfo: ResourceInfo,
+    private val identityServiceClient: IdentityServiceClient,
+    private val jwtValidator: JwtValidator,
+    @ConfigProperty(name = "app.security.access-token-cookie-name", defaultValue = "access_token")
+    private val accessTokenCookieName: String,
+    @ConfigProperty(name = "app.security.refresh-token-cookie-name", defaultValue = "refresh_token")
+    private val refreshTokenCookieName: String
+) : ContainerRequestFilter {
+
+    override fun filter(requestContext: ContainerRequestContext) {
+        val annotation = getAuthenticationAnnotation(resourceInfo) ?: return
+        val accessToken = requestContext.cookies[accessTokenCookieName]?.value
+        val refreshToken = requestContext.cookies[refreshTokenCookieName]?.value
+
+        // Deny if refresh token is missing or invalid
+        if (refreshToken.isNullOrBlank() || !jwtValidator.isTokenValidAndNotBlank(refreshToken)) {
+            denyRequest(requestContext, annotation.message)
+        }
+
+        // If access token is valid, use it; otherwise, attempt to refresh
+        val tokenToUse = when {
+            jwtValidator.isTokenValidAndNotBlank(accessToken) -> accessToken!!
+            refreshToken.isNullOrBlank() -> {
+                denyRequest(requestContext, "Missing refresh token")
+                return
+            }
+            else -> tryRefreshAccessToken(requestContext, refreshToken)
+        }
+
+        // Check if the token has the required role
+        if (!jwtValidator.hasRequiredRole(tokenToUse, annotation.roles.toSet())) {
+            denyRequest(requestContext, annotation.message)
+        }
+    }
+
+    private fun tryRefreshAccessToken(
+        requestContext: ContainerRequestContext,
+        refreshToken: String
+    ): String = runCatching {
+        val newToken = identityServiceClient.refreshAccessToken(refreshToken)
+        val newAccessTokenCookie = createNewAccessTokenCookie(newToken)
+        requestContext.headers.add(HttpHeaders.SET_COOKIE, newAccessTokenCookie.toString())
+        newToken
+    }.getOrElse {
+        denyRequest(requestContext, "Unauthorized")
+        throw UnauthorizedException("Failed to refresh token")
+    }
+
+    private fun createNewAccessTokenCookie(newAccessToken: String): NewCookie =
+        NewCookie.Builder(accessTokenCookieName)
+            .value(newAccessToken)
+            .path("/")
+            .httpOnly(true)
+            .secure(true)
+            .build()
+
+    private fun denyRequest(requestContext: ContainerRequestContext, message: String) {
+        requestContext.abortWith(
+            Response.status(Response.Status.UNAUTHORIZED)
+                .entity(message)
+                .build()
+        )
+    }
```

Comment:
Als ik het goed begrijp stuurt dit een 401 Unauthorized terug met een message:
```
401 Unauthorized
{
   "message": "the provided message"
}
```
Dit volgt helaas niet de `APIException` of `createResponse` structure. Je wilt deze structuur aanhouden. Daarbovenop, als jij een Exception gooit en niet vangt vanuit de `AuthenticationFilter`, wordt die dan opgevangen door de `GlobalExceptionHandler`? Pls test dit want als dit zo is kan je gwn een Exception gooien ipv aborten. 

In kort, zorg ervoor dat dit clean en gestandaardiseerd is met zo min mogelijk duplicate code.